test/testSetup.ts

```diff
@@ -109,6 +124,11 @@ export class TestSetup {
   }
 
   public async createProject(scratchOrgEdition: string): Promise<void> {
+    if (this.usingExistingProject) {
```

Comment:
testSetup is predicated on a single use case, our current e2e test needs. I see the entire repo as a possibility for supporting not just e2e for vsce, but others that need to test extensions of there own that are to use in conjunction with dx projects.

The parallel I am drawing is to [CLI Plugins Testkit](https://github.com/salesforcecli/cli-plugins-testkit/blob/HEAD/SAMPLES.md), which provides much of the same functionality, but focused on plugin development.

I see this repo as the same offering to extension authors.

I agree that the code I have added here is orthogonal to the meaning of the function name, but I see similar issues with dev hub auth. No check exists to confirm that the env var for hub identity are set.

I will move the guard into setup method, bypassing the call to createProject.

Reply:
_"...I see the entire repo as a possibility for supporting not just e2e for vsce, but others that need to test extensions of there own that are to use in conjunction with dx projects"_

The repo _is specifically_ for e2e testing and was not designed to be a generic testing platform.  If we can add (with little overhead and minor changes) support for perf testing, I'm all for that.  However, I do not want to design it for something where a specific requirement and a specific goal do not exist.

At the moment, in apexLspPerf.ts...
`testSetup.setUpTestingEnvironment()` is called
and `testSetup.tearDown()` is also called...
...however, `tearDown()` runs "sfdx force:org:delete" on the scratch org that was created, but in this suite, a scratch org is not created, so `tearDown()` should not be called.

...so
1. The majority of TestSetup is not used by apexLspPerf.ts
2. The only function that's used is `setUpTestingEnvironment()`
3. The majority of the proposed changes do not seem appropriate for this class
4. We're trying to make TestSetup do something it was not meant or designed to do
5. The proposed changes break (IMHO) the single responsibility of the class.

I've been thinking about this, and here's what I suggest:
1. back out the changes to testSetup.ts
2. then, in testSetup.ts, add:
```
  public projectExists(): boolean {
    return fs.existsSync(this.projectFolderPath);
  }
```
...and in `public async setUpTestingEnvironment()`, at line 94 of the existing version, change
```
if (fs.existsSync(this.projectFolderPath)) {
```
...to:
```
if (this.projectExists()) {
```

4. create a new file, perfSetup.ts and have it extend testSetup.ts, implement its own setUpTestingEnvironment() and its own projectExists():
```

export class PerfSetup extends TestSetup {
  public projectFolderPath: string | undefined = undefined;
  public usingExistingProject: boolean;

  public constructor(
    testSuiteSuffixName: string,
    reuseScratchOrg: boolean,
    existingProjectPath?: string
  ) {
    super(testSuiteSuffixName, reuseScratchOrg);

    if (existingProjectPath && !fs.existsSync(existingProjectPath)) {
      throw new Error(
        `In TestSetup(), the existing project path '${existingProjectPath}' does not exist`
      );
    }

    this.usingExistingProject = existingProjectPath !== undefined;
    this.projectFolderPath = existingProjectPath;
  }

  public async setUpTestingEnvironment(): Promise<void> {
    // set projectFolderPath if it is undefined
    this.projectFolderPath ??=
      path.join(this.tempFolderPath, this.tempProjectName);

    utilities.log(
      this.usingExistingProject
        ? `${this.testSuiteSuffixName} - using existing project at ${this.projectFolderPath}`
        : `${this.testSuiteSuffixName} - creating project files in ${this.projectFolderPath}`
    );

    super.setUpTestingEnvironment();
  }

  public projectExists(): boolean {
    return fs.existsSync(this.projectFolderPath) && !this.usingExistingProject;
  }
}
```

This way, perfSetup can use the parts of testSetup it wants, it can customize what it wants, and testSetup remains mostly unaffected.


Or, alternatively, we could create a new project and repo specifically for perf testing.  Creating a new wdio-vscode-service project is pretty easy - just run `npm create wdio ./`

See https://www.npmjs.com/package/wdio-vscode-service for more info on this.

Reply:
I am not sure what the contention is. TestSetup already accounts for runtime defined dx projects; via createProject (or, now, an existing one).

The goal these changes, for my current work, is to support existing projects, not specifically perf testing. With this one change others can now use the framework to use other projects, which may come with other needs.

For perf testing I don't need to create a project, create scratch orgs or auth dev hubs.

I am trying to make TestSetup a bit more flexible w/o breaking how its being used now, by adding some guardrails in places folks might run into issues.

I don't see the need to extend TestSetup move the guardrails out of TestSetup. Composition over inheritance is the pattern I prefer and the subclass is substantially different to, IMO, warrant a subclass.